mk/build_optee_for_android.mk

```diff
@@ -0,0 +1,88 @@
+##########################################################
+## Common mk file used for Android to compile and       ##
+## integrate optee related components                   ##
+## Following variant need to be defined in device.mk    ##
+##    OPTEE_OS_DIR                                      ##
+##    OPTEE_TA_TARGETS                                  ##
+##    OPTEE_CFG_ARM64_CORE                              ##
+##    OPTEE_PLATFORM                                    ##
+## And  BUILD_OPTEE_MK need to be defined in device.mk  ##
+## to point to this file                                ##
+##                                                      ##
+## local_module need to be defined before include this  ##
+## this file to build TAs                               ##
+##                                                      ##
+##########################################################
+
+##########################################################
+## define common variants, like TA_DEV_KIT_DIR          ##
+##########################################################
+OPTEE_OUT_DIR ?= $(PRODUCT_OUT)/optee
+OPTEE_TA_OUT_DIR ?= $(OPTEE_OUT_DIR)/ta
+# Set so that OP-TEE clients can find the installed dev-kit, which
+# depends on platform and its optee word-size.
+OPTEE_OS_OUT_DIR ?= $(OPTEE_OUT_DIR)/arm-plat-$(OPTEE_PLATFORM)
+TA_DEV_KIT_DIR := $(OPTEE_OS_OUT_DIR)/export-${OPTEE_TA_TARGETS}
+
+TOP_ROOT_ABS := $(realpath $(TOP))
+CROSS_COMPILE64 := $(TOP_ROOT_ABS)/$(TARGET_TOOLS_PREFIX)
+CROSS_COMPILE_LINE := CROSS_COMPILE64="$(CROSS_COMPILE64)"
+ifeq ($(strip $($(combo_2nd_arch_prefix)TARGET_TOOLS_PREFIX)),)
+CROSS_COMPILE32 := $(TOP_ROOT_ABS)/$($(combo_2nd_arch_prefix)TARGET_TOOLS_PREFIX)
+CROSS_COMPILE_LINE += CROSS_COMPILE32="$(CROSS_COMPILE32)"
+endif
+
+##########################################################
+## define BUILD_OPTEE_OS target, add condition check    ##
+## to make it only be defined once even this file       ##
+## included multiple times                              ##
+##########################################################
+ifneq (true,$(BUILD_OPTEE_OS_DEFINED))
+BUILD_OPTEE_OS_DEFINED := true
+
+.PHONY: BUILD_OPTEE_OS
+BUILD_OPTEE_OS:
+	@echo "Start building optee_os..."
+	$(MAKE) -C $(TOP_ROOT_ABS)/$(OPTEE_OS_DIR) \
+		O=$(TOP_ROOT_ABS)/$(OPTEE_OS_OUT_DIR) \
+		ta-targets=$(OPTEE_TA_TARGETS) \
+		CFG_ARM64_core=$(OPTEE_CFG_ARM64_CORE) \
+		PLATFORM=$(OPTEE_PLATFORM) \
+		$(CROSS_COMPILE_LINE)
+	@echo "Finished building optee_os..."
+
+endif
+
+##########################################################
+## Lines for building TAs automatically                 ##
```

Comment:
I'm not sure what you mean by "building TAs automatically". The modules are tagged as optional. They should be included always and if we need to build them, they need to be added explicitly in the device folder to the PRODUCT_PACKAGES.